codec-http2/src/test/java/io/netty/handler/codec/http2/DefaultHttp2FrameReaderTest.java

```diff
@@ -386,6 +386,40 @@ public void execute() throws Throwable {
         }
     }
 
+    @Test
+    public void verifyValidRequestAfterMalformedPacketCausesStreamException() throws Http2Exception {
+        final ByteBuf input = Unpooled.buffer();
+        int priorityStreamId = 3, headerStreamId = 5;
+        try {
+            // Write a malformed priority header causing a stream exception in reader
+            writeFrameHeader(input, 4, PRIORITY, new Http2Flags(), priorityStreamId);
+            assertThrows(Http2Exception.class, new Executable() {
+                @Override
+                public void execute() throws Throwable {
+                    try {
+                        frameReader.readFrame(ctx, input, listener);
+                    } catch (Exception e) {
+                        if (e instanceof Http2Exception && Http2Exception.isStreamError((Http2Exception) e)) {
+                            throw e;
+                        }
+                    }
+                }
+            });
+            // Verify that after stream exception we accept new stream requests
```

Comment:
@isaacrivriv I read description of #13788 to understand expected behavior, but it seems like this is not what RFC wants. [Section 4.2](https://datatracker.ietf.org/doc/html/rfc9113#section-4.2) says:

> An endpoint MUST send an error code of [FRAME_SIZE_ERROR](https://datatracker.ietf.org/doc/html/rfc9113#FRAME_SIZE_ERROR) if a frame exceeds the size defined in [SETTINGS_MAX_FRAME_SIZE](https://datatracker.ietf.org/doc/html/rfc9113#SETTINGS_MAX_FRAME_SIZE), exceeds any limit defined for the frame type, or is too small to contain mandatory frame data. A frame size error in a frame that could alter the state of the entire connection MUST be treated as a [connection error](https://datatracker.ietf.org/doc/html/rfc9113#ConnectionErrorHandler) ([Section 5.4.1](https://datatracker.ietf.org/doc/html/rfc9113#ConnectionErrorHandler))


Then section 5.4.1 describes that the state of the entire connection is expected to be corrupted at that point and we must close it. It's not expected that we could recover and process more frames on the same connection.

If we don't already have one, it will be helpful if you could implement an integration test with a full HTTP/2 server and a faked client that send a malformed frame, then assert that the server responds with `FRAME_SIZE_ERROR`, then closes connection.

Reply:
In this case, the behavior I'm seeing with the response from a Netty HTTP2 server matches the RFC specification where in some cases even though a malformed frame comes in, it doesn't actually shut down the connection but just resets a stream and more streams can be accepted after that. [Section 6.3](https://datatracker.ietf.org/doc/html/rfc9113#name-priority) is an example

> A PRIORITY frame with a length other than 5 octets MUST be treated as a [stream error](https://datatracker.ietf.org/doc/html/rfc9113#StreamErrorHandler) ([Section 5.4.2](https://datatracker.ietf.org/doc/html/rfc9113#StreamErrorHandler)) of type [FRAME_SIZE_ERROR](https://datatracker.ietf.org/doc/html/rfc9113#FRAME_SIZE_ERROR).

As it is currently working if I send a priority frame with length 4, I correctly get a reset frame back with a frame size error (which is what the RFC states) but starting another stream with a valid request after that, it is not parsed correctly due to leftover data from the previous request.

Reply:
The most important part is: 

> A frame size error in a frame that could alter the state of the entire connection MUST be treated as a [connection error](https://datatracker.ietf.org/doc/html/rfc9113#ConnectionErrorHandler) ([Section 5.4.1](https://datatracker.ietf.org/doc/html/rfc9113#ConnectionErrorHandler))

It's ok to send `RST_STREAM` frame first. But then, how can we trust the frame if it has incorrect format? If we can not trust it, we don't know how many bytes to skip. It could happen that we received total garbage that just happen to start with "priority frame header". So, if we don't know how to continue, we should send GO_AWAY and close connection.

Reply:
The RFC states the following after that statement which confuses me

> A frame size error in a frame that could alter the state of the entire connection MUST be treated as a [connection error](https://datatracker.ietf.org/doc/html/rfc9113#ConnectionErrorHandler) ([Section 5.4.1](https://datatracker.ietf.org/doc/html/rfc9113#ConnectionErrorHandler)); **this includes any frame carrying a [field block](https://datatracker.ietf.org/doc/html/rfc9113#FieldBlock) ([Section 4.3](https://datatracker.ietf.org/doc/html/rfc9113#FieldBlock)) (that is, [HEADERS](https://datatracker.ietf.org/doc/html/rfc9113#HEADERS), [PUSH_PROMISE](https://datatracker.ietf.org/doc/html/rfc9113#PUSH_PROMISE), and [CONTINUATION](https://datatracker.ietf.org/doc/html/rfc9113#CONTINUATION)), a [SETTINGS](https://datatracker.ietf.org/doc/html/rfc9113#SETTINGS) frame, and any frame with a stream identifier of 0.**

So in my head the RFC states the specifics of what "altering the state of the connection" is in that sentence which this particular case doesn't seem to apply. By sending a priority frame to a specific stream (not 0) we can hit this issue and doesn't really fall under those categories IMO. 

I think we should skip the bytes if the payload length matches the actual frame payload and send a reset stream which matches the RFC and continue reading for the cases where a stream reset is required. I also agree closing the connection on cases where the payload length doesn't actually match the frame payload considering that we are receiving garbage in those cases but I'm making sure this isn't against the RFC since it states otherwise about keeping the connection active.

The other question would be, how would we know if the frame that came in actually matches the payload length? By the time Netty start's parsing H2 data we could have more messages queued in the buffer which we're not aware of. I'm thinking maybe a check on every decode to peek at the payload length and compare to the actual frame length? I'm not aware of a way to cleanly distinguish frames from each other TBH.

Reply:
See my comment above but the TL;DR is that the length field from the frame header is authoritative on size of a frames payload. What to do if that payload size is too short or too long when decoding the frame is frame type specific and is described as part of the individual frame definitions.

Reply:
@normanmaurer, I am. This should prevent frame stream corruption for problems that result in a stream error at the validation phase.

One behavior change that I think is worth point out is that we'll now aggregate the data for the next frame before doing any of the per-frame validations other than max frame length. I think this is safe because we've already agreed to accept frames that big such as data and headers frames but it's worth a thought.

Reply:
@isaacrivriv thanks for pointing to the continuation of that sentence. That makes sense, for the current case we don't need to close the entire connection. Great catch!